pmlc/rt/vulkan/wrappers.cc

```diff
@@ -21,111 +21,207 @@
 #include "pmlc/rt/vulkan/vulkan_runtime.h"
 
 namespace {
-
 class VulkanRuntimeManager {
 public:
-  VulkanRuntimeManager() = default;
+  VulkanRuntimeManager() {
+    if (failed(vulkanRuntime.init())) {
+      llvm::errs() << "vulkanRuntime.init() failed";
+    }
+  }
   VulkanRuntimeManager(const VulkanRuntimeManager &) = delete;
   VulkanRuntimeManager operator=(const VulkanRuntimeManager &) = delete;
-  ~VulkanRuntimeManager() = default;
-
-  void setResourceData(DescriptorSetIndex setIndex, BindingIndex bindIndex,
-                       const VulkanHostMemoryBuffer &memBuffer) {
-    std::lock_guard<std::mutex> lock(mutex);
-    vulkanRuntime.setResourceData(setIndex, bindIndex, memBuffer);
+  ~VulkanRuntimeManager() {
+    if (failed(vulkanRuntime.destroy())) {
+      llvm::errs() << "vulkanRuntime.destroy() failed";
+    }
   }
 
-  void setEntryPoint(const char *entryPoint) {
+  void createLaunchKernelAction(uint8_t *shader, uint32_t size,
+                                const char *entryPoint,
+                                NumWorkGroups numWorkGroups) {
     std::lock_guard<std::mutex> lock(mutex);
+    vulkanRuntime.createLaunchKernelAction();
+    vulkanRuntime.setShaderModule(shader, size);
     vulkanRuntime.setEntryPoint(entryPoint);
+    vulkanRuntime.setNumWorkGroups(numWorkGroups);
   }
 
-  void setNumWorkGroups(NumWorkGroups numWorkGroups) {
+  void createMemoryTransferAction(uint64_t src_index, uint64_t src_binding,
+                                  uint64_t dst_index, uint64_t dst_binding) {
     std::lock_guard<std::mutex> lock(mutex);
-    vulkanRuntime.setNumWorkGroups(numWorkGroups);
+    vulkanRuntime.createMemoryTransferAction(src_index, src_binding, dst_index,
+                                             dst_binding);
   }
 
-  void setShaderModule(uint8_t *shader, uint32_t size) {
+  void setResourceData(DescriptorSetIndex setIndex, BindingIndex bindIndex,
+                       const VulkanHostMemoryBuffer &memBuffer) {
     std::lock_guard<std::mutex> lock(mutex);
-    vulkanRuntime.setShaderModule(shader, size);
+    vulkanRuntime.setResourceData(setIndex, bindIndex, memBuffer);
   }
 
-  void runOnVulkan() {
+  void setLaunchKernelAction() {
     std::lock_guard<std::mutex> lock(mutex);
-    if (failed(vulkanRuntime.initRuntime()) || failed(vulkanRuntime.run()) ||
-        failed(vulkanRuntime.updateHostMemoryBuffers()) ||
-        failed(vulkanRuntime.destroy())) {
+    if (failed(vulkanRuntime.setLaunchKernelAction())) {
       llvm::errs() << "runOnVulkan failed";
     }
   }
 
+  void addLaunchActionToSchedule() {
+    std::lock_guard<std::mutex> lock(mutex);
```

Comment:
What's the thinking behind the locking? Do we anticipate this to occur from multiple threads? It seems like the host-function we generate will be executed on a single thread.

Reply:
This lock is tend to avoid possible action disorder in the future. Yes, it should run on a single thread now and we can safely remove it. Will we consider multiple threads sharing the same schedule and generating actions concurrently in the future?

Reply:
I don't think so since generating the schedule should only happen once during the `init` function eventually.